source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -211,8 +211,14 @@ Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::ResponseTrailerMa
     return Http::HeaderMap::Iterate::Continue;
   });
 
-  // Clear out the trailers so they don't get added since it is now in the body
-  trailers.clear();
+  // Due to https://github.com/envoyproxy/envoy/issues/10514 (and
+  // https://bugs.webkit.org/show_bug.cgi?id=210108) we need to skip clearing response trailer
+  // when the downstream protocol is HTTP/2.
+  if (encoder_callbacks_->streamInfo().protocol() != Http::Protocol::Http2) {
```

Comment:
Can we use more specific workaround for this instead of sending trailers to all H2 client? e.g. inspecting user-agent or something else?

Reply:
This bug also effects Internet Explorer 11￼ and Edge on various patch versions of Windows 10 (pre Chromium); so inspecting user-agents could miss some effected browsers.

Reply:
Hum, need to update the description and comments then?

Reply:
How about clearing the trailers when H2 and Chrome/Chromium only? Yeap. user-agent detection is tricky.

Reply:
What is the downside of sending the trailers (I'm presuming twice as they are in the body too) to all H2 clients for grpc-web? 
Doing any kind of browser filtering could lead to future debugging headaches.

Reply:
On a second thought, I don't think gRPC-web need a response trailers ever, so shall we always clear (or say, not sending HEADERS frame at the end in case of H2)? I don't think in H1 we send trailers by default so it's not emit anyway.

Reply:
Agreed. For grpc-web, we should always drop the trailers as the data has been copied to the body.

===

Is it correct that the general problem could also apply to HTTP endpoints which choose to emit any trailers,  if Envoy or any forward proxies choose to forward trailers to a browser client? If so, it seems that we should make trailer forwarding an opt-in config for both H1 and H2 in Envoy.

Reply:
Agree that we should not need the trailers for gRPC-web, but that is the current behaviour that has caused the described issue in Safari and IE. By sending the trailers when H2 resolves the problem; but at the cost of sending the trailers twice

Reply:
@lizan a good catch. I think we can skip submitting trailers when it is cleared. Let me test it.